File: python_backend/app/domain/entities/plan.py
"""Plan domain entity (B5 — Billing).

Pure data — no framework or infrastructure dependencies. ``limits`` is a free-form
dict so we can evolve tier perks (max minutes, max docs, feature flags) without
schema changes. ``price_cents`` is integer cents per NFR-22 (no floating point).
"""
from dataclasses import dataclass, field
from typing import Any, Literal, Optional

PlanCode = Literal["free", "pro", "premium", "byok"]
BillingCycle = Literal["free", "monthly", "yearly", "lifetime"]

VALID_PLAN_CODES: set[str] = {"free", "pro", "premium", "byok"}
VALID_BILLING_CYCLES: set[str] = {"free", "monthly", "yearly", "lifetime"}


@dataclass
class Plan:
    """A plan in the catalog (free/pro_monthly/pro_yearly/etc.).

    Plans are seeded via ``seed_plans`` on init_db; the table acts as a
    catalog so pricing changes don't require a deploy. ``is_legacy`` flags
    plans that are no longer offered to new users but still honored for
    existing subscribers (grandfathering — NFR-11/NFR-14)."""

    id: str
    code: PlanCode
    name: str
    description: Optional[str]
    price_cents: int
    currency: str
    billing_cycle: BillingCycle
    lemon_squeezy_variant_id: Optional[str]
    lemon_squeezy_product_id: Optional[str]
    limits: dict[str, Any] = field(default_factory=dict)
    is_active: bool = True
    is_legacy: bool = False
    sort_order: int = 0

File: apps/web/src/domain/entities/plan.ts
/**
 * Plan domain entity — TypeScript mirror of the backend `Plan` entity.
 *
 * Backend contract (python_backend/app/domain/entities/plan.py exposed via
 * GET /api/plans as `PlanResponse`):
 *
 *   {
 *     id: string,                    // e.g. "free", "pro_monthly", ...
 *     code: "free" | "pro" | "premium" | "byok",
 *     name: string,
 *     description: string | null,
 *     price_cents: integer,
 *     currency: string,              // "USD"
 *     billing_cycle: "free" | "monthly" | "yearly" | "lifetime",
 *     limits: {
 *       max_session_duration_minutes: number | null,  // null = unlimited
 *       max_minutes_per_month: number | null,
 *       max_docs: number | null,
 *       max_recordings: number | null,                // 0 = blocked
 *       max_storage_gb: number,
 *       max_share_links: number | null,
 *       max_custom_scenarios: number | null,
 *       diarization_enabled: boolean,
 *       voice_fingerprinting_enabled: boolean,
 *       tweaks_layouts_unlocked: string[],
 *       tweaks_hint_styles_unlocked: string[],
 *       tweaks_transcript_styles_unlocked: string[],
 *       byok_enabled: boolean,
 *       priority_support: boolean,
 *       export_formats: string[],
 *       stealth_mode: boolean
 *     },
 *     is_active: boolean,
 *     is_legacy: boolean,
 *     sort_order: number
 *   }
 *
 * Frontend uses camelCase + integer cents (money) and re-exposes
 * `limits` in snake_case to match the backend `PlanLimits` shape.
 */

export type PlanCode = "free" | "pro" | "premium" | "byok";
export type BillingCycle = "free" | "monthly" | "yearly" | "lifetime";

export interface PlanLimits {
  max_session_duration_minutes: number | null;
  max_minutes_per_month: number | null;
  max_docs: number | null;
  max_recordings: number | null;
  max_storage_gb: number;
  max_share_links: number | null;
  max_custom_scenarios: number | null;
  diarization_enabled: boolean;
  voice_fingerprinting_enabled: boolean;
  tweaks_layouts_unlocked: string[];
  tweaks_hint_styles_unlocked: string[];
  tweaks_transcript_styles_unlocked: string[];
  byok_enabled: boolean;
  priority_support: boolean;
  export_formats: string[];
  stealth_mode: boolean;
}

export interface Plan {
  id: string;
  code: PlanCode;
  name: string;
  description: string | null;
  priceCents: number;
  currency: string;
  billingCycle: BillingCycle;
  limits: PlanLimits;
  isActive: boolean;
  isLegacy: boolean;
  sortOrder: number;
}
